pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/OrDocIdIterator.java

```diff
@@ -15,139 +15,119 @@
-      currentDocId = queue.peek().getLeft();
+    if (nextDocId > _maxDocId) {
+      _currentDocId = Constants.EOF;
     } else {
-      currentDocId = Constants.EOF;
+      _currentDocId = nextDocId;
+      if (hasExhaustedIterator) {
+        removeExhaustedIterators();
+      }
     }
-
-    long end = System.nanoTime();
-    timeMeasure.addAndGet(end - start);
-    return currentDocId;
+    return _currentDocId;
   }
 
   @Override
-  public int next() {
-    long start = System.currentTimeMillis();
-
-    if (currentDocId == Constants.EOF) {
-      return currentDocId;
+  public int advance(int targetDocId) {
+    if (_currentDocId == Constants.EOF) {
+      return Constants.EOF;
     }
-    while (queue.size() > 0 && queue.peek().getLeft() <= currentDocId) {
-      IntPair pair = queue.remove();
-      iteratorIsInQueue[pair.getRight()] = false;
+    if (targetDocId > _maxDocId) {
+      _currentDocId = Constants.EOF;
+      return Constants.EOF;
     }
-    currentDocId++;
-    // Grab the next value from each iterator, if it's not in the queue
-    for (int i = 0; i < docIdIterators.length; i++) {
-      if (!iteratorIsInQueue[i]) {
-        int nextDocId = docIdIterators[i].advance(currentDocId);
-        if (nextDocId != Constants.EOF) {
-          if (!(nextDocId <= maxDocId && nextDocId >= minDocId) && nextDocId >= currentDocId) {
-            throw new RuntimeException("next Doc : " + nextDocId
-                + " should never crossing the range : [ " + minDocId + ", " + maxDocId + " ]");
-          }
-          queue.add(new IntPair(nextDocId, i));
-        }
-        iteratorIsInQueue[i] = true;
-      }
+    if (targetDocId <= _currentDocId) {
+      return _currentDocId;
+    }
+
+    if (targetDocId < _minDocId) {
+      targetDocId = _minDocId;
     }
 
-    if (queue.size() > 0) {
-      currentDocId = queue.peek().getLeft();
+    int nextDocId = Integer.MAX_VALUE;
+    boolean hasExhaustedIterator = false;
+    for (int i = 0; i < _numNotExhaustedIterators; i++) {
+      int docId = _nextDocIds[i];
+      if (docId < targetDocId) {
+        docId = _docIdIterators[i].advance(targetDocId);
+        _nextDocIds[i] = docId;
+      }
+      if (docId != Constants.EOF) {
+        nextDocId = Math.min(nextDocId, docId);
+      } else {
+        hasExhaustedIterator = true;
+      }
+    }
+    if (nextDocId > _maxDocId) {
+      _currentDocId = Constants.EOF;
     } else {
-      currentDocId = Constants.EOF;
+      _currentDocId = nextDocId;
+      if (hasExhaustedIterator) {
+        removeExhaustedIterators();
+      }
     }
-    long end = System.currentTimeMillis();
-    timeMeasure.addAndGet(end - start);
-    // Remove this after tracing is added
-    // if (currentDocId == Constants.EOF) {
-    // LOGGER.debug("OR took:" + timeMeasure.get());
-    // }
+    return _currentDocId;
+  }
 
-    return currentDocId;
+  private void removeExhaustedIterators() {
+    int i = 0;
+    while (i < _numNotExhaustedIterators) {
```

Comment:
this probably wont work if two iterators exhaust in one round right? lets say the first and last iterator exhaust, we will copy the last entry to the first position in the array but that would  be pointing to EOF.

Another case is if the last iterator exhausts, we copy the same value to the same position.

see if you can you simplify the logic by maintaining a boolean array for iteratorState.

Reply:
It works. Will add comment to explain